src/compiler/scala/tools/nsc/transform/Delambdafy.scala

```diff
@@ -347,7 +347,7 @@ abstract class Delambdafy extends Transform with TypingTransformers with ast.Tre
     private val thisReferringMethods = mutable.Set[Symbol]()
 
     // the set of lifted lambda body methods that each method refers to
-    private val liftedMethodReferences = mutable.Map[Symbol, Set[Symbol]]().withDefault(_ => mutable.Set())
+    private val liftedMethodReferences = mutable.Map[Symbol, mutable.Set[Symbol]]()
```

Comment:
No more `withDefault`?

Reply:
See the changes a few lines further down:

```
-        if (currentMethod.exists) liftedMethodReferences(currentMethod) += sel.symbol
+        if (currentMethod.exists) liftedMethodReferences.getOrElseUpdate(currentMethod, mutable.Set()) += sel.symbol
```

This did the right thing only by accident with the old collections (but doesn't anymore with the new ones). The intention is apparently to update the mutable set but a default is computed from scratch on every access, it is never stored back into the map automatically, i.e. `liftedMethodReferences(currentMethod) += sel.symbol` computes an empty mutable set as the default value, adds the symbol to it, and then discards the result. It worked with the old collections because `+=` actually desugared to a `+` for the mutable set (thus computing a new mutable set instead of updating in place) and an `update_=` for the map.

We still have `withDefault` and could use it here (while fixing only the calls to update the sets and map) but not doing so makes the code clearer.

Reply:
I agree the old code was unclear (trying to be too clever). I'm curious to understand why it stopped working in the new collections. Is it because there's a `+=` method on immutable `Set` now? Or what is the precise change that broke this?

Reply:
Note that it's only typed as `collection.Set`, not `mutable.Set`. We removed `collection.Set.+` to prevent users from accidentally calling it on mutable sets. Since the code looked like it should call `+=` the obvious change that I did initially was to change the type to `mutable.Set` which led to the bug because it didn't really call `+=` before.

We have since added back a deprecated version of `collection.Set.+` so that the original code would compile (and work the same way it did in 2.12).

Reply:
Cool, thanks for explaining!